In [ ]:
import pandas as ps
import numpy as np
import seaborn as sb


from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsRegressor

datos = ps.read_csv('/content/sample_data/diabetes_exa.csv', delimiter=',')
datos


preg  plas  pres  skin  insu  mass   pedi  age     class
0       5   116    74    29     0  32.3  0.660   35  positive
1       0   198    66    32   274  41.3  0.502   28  positive
2       4    90     0     0     0  28.0  0.610   31  negative
3       8   167   106    46   231  37.6  0.165   43  positive
4       8   155    62    26   495  34.0  0.543   46  positive
..    ...   ...   ...   ...   ...   ...    ...  ...       ...
743     5   155    84    44   545  38.7  0.619   34  negative
744     3   111    62     0     0  22.6  0.142   21  negative
745     1   111    86    19     0  30.1  0.143   23  negative
746     5    73    60     0     0  26.8  0.268   27  negative
747     1    80    55     0     0  19.1  0.258   21  negative

[748 rows x 9 columns]

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
print(datos.groupby('class').size()) #obteniendo la medida de la clasificacion de resultados de pacientes con diabetes

class
negative    486
positive    262
dtype: int64


In [ ]:
X_train, X_test = train_test_split(datos, test_size=0.3, random_state=42)
X_train

preg  plas  pres  skin  insu  mass   pedi  age     class
369     0   108    68    20     0  27.3  0.787   32  negative
628     8    65    72    23     0  32.0  0.600   42  negative
235     1   153    82    42   485  40.6  0.687   23  negative
585     6    80    80    36     0  39.8  0.177   28  negative
209     1   168    88    29     0  35.0  0.905   52  positive
..    ...   ...   ...   ...   ...   ...    ...  ...       ...
71      8   120     0     0     0  30.0  0.183   38  positive
106    13   106    72    54     0  36.6  0.178   45  negative
270     9    72    78    25     0  31.6  0.280   38  negative
435    10   179    70     0     0  35.1  0.200   37  negative
102     0   127    80    37   210  36.3  0.804   23  negative

[523 rows x 9 columns]

In [ ]:
y_train = X_train['class']
y_test = X_test['class']

In [ ]:
X_train = X_train.drop(['class'],axis=1)
X_train

preg  plas  pres  skin  insu  mass   pedi  age
369     0   108    68    20     0  27.3  0.787   32
628     8    65    72    23     0  32.0  0.600   42
235     1   153    82    42   485  40.6  0.687   23
585     6    80    80    36     0  39.8  0.177   28
209     1   168    88    29     0  35.0  0.905   52
..    ...   ...   ...   ...   ...   ...    ...  ...
71      8   120     0     0     0  30.0  0.183   38
106    13   106    72    54     0  36.6  0.178   45
270     9    72    78    25     0  31.6  0.280   38
435    10   179    70     0     0  35.1  0.200   37
102     0   127    80    37   210  36.3  0.804   23

[523 rows x 8 columns]

In [ ]:
X_test = X_test.drop(['class'],axis=1)
X_test

preg  plas  pres  skin  insu  mass   pedi  age
580     0   123    72     0     0  36.3  0.258   52
356     1   140    74    26   180  24.1  0.828   23
133     7   195    70    33   145  25.1  0.163   55
250     9    89    62     0     0  22.5  0.142   33
299     4   145    82    18     0  32.5  0.235   70
..    ...   ...   ...   ...   ...   ...    ...  ...
182     4   117    62    12     0  29.7  0.380   30
567     2    88    74    19    53  29.0  0.229   22
264     4    83    86    19     0  29.3  0.317   34
434     9   124    70    33   402  35.4  0.282   34
223     4   111    72    47   207  37.1  1.390   56

[225 rows x 8 columns]

In [ ]:
ct = GaussianNB()

a) Obtén un clasificador en Python para la condición de Diabetes.

R= El clasificador es GaussianNB


In [ ]:
ct.fit(X_train,y_train)

GaussianNB()

In [ ]:
#se obtiene el pronostico de X_test
pronostico = ct.predict(X_test)

In [ ]:
pronostico

array(['negative', 'negative', 'positive', 'negative', 'negative',
       'negative', 'negative', 'positive', 'positive', 'positive',
       'negative', 'negative', 'negative', 'negative', 'negative',
       'negative', 'positive', 'negative', 'positive', 'negative',
       'negative', 'negative', 'negative', 'positive', 'negative',
       'negative', 'positive', 'negative', 'negative', 'positive',
       'negative', 'negative', 'negative', 'negative', 'negative',
       'negative', 'negative', 'negative', 'negative', 'negative',
       'negative', 'negative', 'negative', 'negative', 'negative',
       'negative', 'negative', 'negative', 'negative', 'positive',
       'positive', 'negative', 'negative', 'negative', 'positive',
       'positive', 'negative', 'negative', 'negative', 'negative',
       'negative', 'positive', 'positive', 'positive', 'negative',
       'positive', 'negative', 'positive', 'negative', 'positive',
       'negative', 'positive', 'negative', 'negative', 'negati

c) ¿Cuál es la eficiencia del modelo que obtuviste?
La eficiencia es del 73%


In [ ]:
#puntaje de exactitud del modelo de clasificación
print (ct.score(X_test,y_test))

0.7333333333333333


d) ¿Cuántos verdaderos positivos y cuántos falsos negativos resultaron?
R= Se tuvieron 118 verdaderos negativos y 35 falsos negativos

In [ ]:
confusion_matrix(y_test,pronostico)

array([[118,  25],
       [ 35,  47]])

In [ ]:
y_test.value_counts()

negative    143
positive     82
Name: class, dtype: int64

In [ ]:
y_test

580    positive
356    negative
133    positive
250    negative
299    positive
         ...   
182    positive
567    negative
264    negative
434    negative
223    positive
Name: class, Length: 225, dtype: object

In [ ]:
datospron = ps.read_csv('/content/sample_data/diabetes_para_pronostico.csv', delimiter=',')
datospron

preg  plas  pres  skin  insu  mass   pedi  age
0      7   133    88    15   155  32.4  0.262   37
1      5    77    82    41    42  35.8  0.156   35
2     12   140    82    43   325  39.2  0.528   58
3      5   106    82    30     0  39.5  0.286   38
4      1    85    66    29     0  26.6  0.351   31
5      0   106    70    37   148  39.4  0.605   22
6      0    94     0     0     0   0.0  0.256   25
7      6   137    61     0     0  24.2  0.151   55
8     11   143    94    33   146  36.6  0.254   51
9      3    88    58    11    54  24.8  0.267   22
10     3   139    54     0     0  25.6  0.402   22
11     6    99    60    19    54  26.9  0.497   32
12     2   105    80    45   191  33.7  0.711   29
13     1    97    66    15   140  23.2  0.487   22
14    12   100    84    33   105  30.0  0.488   46
15     6   162    62     0     0  24.3  0.178   50
16     4   142    86     0     0  44.0  0.645   22
17     2   122    76    27   200  35.9  0.483   26
18     1   143    84    23   310  42.4  1.076   22
19     1   139    62    41   480  40.7  0.536   21

b) Obtén la condición de diabetes de las pacientes del archivo
“datos_para_pronostico.csv”

In [ ]:
prediccion = ct.predict(datospron)
print(prediccion)

['negative' 'negative' 'positive' 'negative' 'negative' 'negative'
 'negative' 'negative' 'positive' 'negative' 'negative' 'negative'
 'negative' 'negative' 'negative' 'negative' 'negative' 'negative'
 'positive' 'positive']


e) i- la Precisión del modelo = (118 + 47) / (118 + 25 + 35 + 47) = 73.3 %

e) ii - la Sensibilidad del modelo. Sensibilidad = VP / (VP + FN)  = 47 / 47+35 = 57.3 %

e) iii - El valor F1.
F1 = 2 *( (Precisión * Sensibilidad) / (Precisión + Sensibilidad) )
F1 = 2 *(0.733 * 0.573)/(0.733+0.573) = 0.643